include/seqan/blast/blast_statistics.h

```diff
@@ -0,0 +1,1076 @@
+
+/*!
+ * @class BlastScoringScheme
+ * @brief Wrapper around a SeqAn scoring scheme
+ * @signature struct BlastScoringScheme<TScoringScheme> { ... };
+ * @headerfile <seqan/blast.h>
+ *
+ * This class wraps around a SeqAn scoring scheme and add functions for the retrievel of pre-computed
+ * Karlin-Altschul-Values (required for e-value statistics) and it modifies the gap-scoring slightly.
+ * If you use high-level IO (@link BlastReport @endlink, @link BlastTabular @endlink) you need not worry about the
+ * specifics, however you should always make sure that the scoring scheme @link BlastScoringScheme#isValid @endlink
+ * after you modify it (or check the return values of the set* functions).
+ *
+ * To retrieve a score-object compatible with regular SeqAn-functions, call
+ * @link BlastScoringScheme#seqanScheme @endlink.
+ *
+ * @section Gap cost computation
+ *
+ * Gap-costs are computed differently in SeqAn and in BLAST.
+ * Blast (and many other tools) compute scores of a stretch of gaps as
+ * <tt>s = gO + n * gE</tt>
+ * where gO is the gapOpen score, gE is the gap extend score and n ist the
+ * total number of gap characters.
+ *
+ * SeqAn, however, normally computes them as as <tt>s = gO + (n-1) * gE</tt>.
+ *
+ * Effectively this results only in a different gap open score. To make dealing with this easier, there are
+ * convenience functions for setting and getting the gap open score in blast convention:
+ * @link BlastScoringScheme#scoreGapOpenBlast @endlink and @link BlastScoringScheme#setScoreGapOpenBlast @endlink.
+ *
+ * @section Karlin-Altschul values
+ *
+ * E-Value statistics require certain pre-computed constants. The BlastScoringScheme picks these automatically
+ * whenever you modify it. Unfortunately there aren't parameters for all combinations of scoring schemes and gap
+ * costs, so you have to check @link BlastScoringScheme#isValid @endlink after modifying it or verify the return
+ * value of the set* functions which is bool here (instead of void).
+ *
+ * More details on the scoring parameters is in these slides:
+ * <a href="https://www.cs.umd.edu/class/fall2011/cmsc858s/Local_Alignment_Statistics.pdf">
+ * https://www.cs.umd.edu/class/fall2011/cmsc858s/Local_Alignment_Statistics.pdf</a>.
+ *
+ * The constants used in SeqAn were imported from the NCBI blast source code (blast-2.2.26)
+ *
+ * @tparam TScoringScheme A SeqAn Score type
+ */
+
+
+/* NOTE(h-2): Maybe this should specialize Score<> instead of being a wrapper.
+ * This would however involve adding a new template parameter to score and
+ * changing lots of code. Also it would become confusing as the shortcuts
+ * like Blosum62 would no longer work...
+ */
+template <typename TScore>
+struct BlastScoringScheme
+{
+    typedef typename Value<TScore>::Type TValue;
+    typedef typename KarlinAltschulValues<TScore, void>::size_t TNumValues;
+
+    /* related to ScoringScheme conversion */
+    TValue _gapOpenBlast = 0;
+    TScore _internalScheme;
```

Comment:
Because of the possibly confusing conversion of open scores I would rather have the people use the set*() functions instead.